Add tests using mocked objects

magicDGS opened the pull request:
As promised in #71

magicDGS commented:
Friendly ping here, @cmnbroad! I would like to have your approval for starting implementing more mocked tests and submit more PRs!

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DocGenBaseTest.java:
```diff
@@ -0,0 +1,76 @@
+    protected DocWorkUnit createDocWorkUnit(final Class<?> clazz) {
+        return createDocWorkUnit(clazz, emptyClassDoc());
+    }
+}
```
I think it might be more useful for this class to something that is instantiated by unit test classes, rather than be a superclass for them. Then the tests could instantiate one of these (maybe passing in whatever doclet or workunit class they want), or just call static methods, and create appropriate mocks for the given test. This would make the tests totally self-contained (i.e., the testPropertyMap test method below would have everything it depends on defined locally, right in the method), and this class could then be used to write unit tests for, say, both the default doclet and the "test" doclet. At some point, we might even want to create another test doclet class to test some feature, and that doclet could also be passed to this class to use to create mocks. Just looking at this it doesn't seem that hard to do, and would be more flexible, but let me know if I'm missing something.

magicDGS commented:
I think that it is a good idea to separate testing and no-testing code, @cmnbroad. Following your suggestions with some modifications, I added three commits to this PR:

* Repackage test input classes: now the documented features/clps for testing live in `org.broadinstitute.barclay.help.testinputs`
* Repackage test doclet code: although you mention to keep the mocking test separate, I rather prefer to separate code not performing any test; that's why the `TestDoclet` and `TestDocWorkUnitHandler` now live in `org.broadinstitute.barclay.help.testdoclets`
* I refactor also the `DocGenBaseTest`to be an utility class for mocking (now called `DocGenMocking`.

---

*Note: I rebased the commits in the latest master to be sure that the changes does not affect the bash-completion code*

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+    @Test(dataProvider = "workUnitForDocStrings")
+    public void testGetDocumentationStrings(final DocWorkUnit workUnit,
+            final String description, final String summary,
+            final String groupName, final String groupSummary) throws Exception {
```
all of these expected values should have names that are prefixed with "expected...".

magicDGS replied:
Done.

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+        return new Object[][] {
+                {DocGenMocking.createDocWorkUnit(DEFAULT_DOC_WORK_UNIT_HANDLER, TestArgumentContainer.class),
+                        "", "Argument container class for testing documentation generation.",
+                        "Test feature group name", "Test program group used for testing"},
```
It took me a little hunting to figure where these strings came from. They should be factored out into symbolic constants (probably defined in the test input class) that can be used both here and in the test input class. Then its easy to see why they're the expected values, and they'll be prefixed with, say "TestArgumentContainer." so they'll be easily navigable in the IDE.

magicDGS replied:
Done.

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getDescription(workUnit), description);
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getSummaryForWorkUnit(workUnit), summary);
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getGroupNameForWorkUnit(workUnit), groupName);
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getGroupSummaryForWorkUnit(workUnit), groupSummary);
```
I think we ultimately want to shoot for a structure with (at least) one unit test per method, and symmetry between the methods and test methods, even if we don't have complete unit test coverage now. This test method should be replaced with 4 separate test methods. That would also cut down on the number of "expected" arguments for each test method.

You may need to produce a separate data provider for each test method (like this test, which uses two work units, one from TestArgumentContainer and one from TestExtraDocs, each with the corresponding expected value for the data entry, but thats ok.

cmnbroad replied:
Also, these unit tests are really helpful to point out flaws in the underlying code - I never fully refactored the code I ported form GATK3 and it shows. All of the methods being tested here are DocWorkUnitHandler methods, but every single one of them takes a workUnit argument, and the implementations only use the work unit itself. These should all be methods on DocWorkUnit! But thats for another day. Still, these kinds of tests are really helpful since they make that easy to see.

magicDGS replied:
Refactor to have one test per method.

cmnbroad commented:
@magicDGS Also, on this:

> I rather prefer to separate code not performing any test; that's why the TestDoclet and TestDocWorkUnitHandler now live in org.broadinstitute.barclay.help.testdoclets

I think its fine for now, but once we start to accumulate more mocks and tests, I may want to separate them as I suggested. But we can hold off on that.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class DefaultDocWorkUnitHandlerTest {
```
Done.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+            new DefaultDocWorkUnitHandler(new HelpDoclet());
+
+    @DataProvider
+    public Object[][] workUnitForDocStrings() throws Exception {
```
I use to add them in every test, in case that methods are changed to throw an exception and it is missing. Removed here.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+                {DocGenMocking.createDocWorkUnit(DEFAULT_DOC_WORK_UNIT_HANDLER, TestExtraDocs.class, DocGenMocking.mockClassDoc("Javadoc description")),
+                        "", "Javadoc description",
+                        "Test extra docs group name", ""},
+                // TODO: because DefaultWorkUnitHandler does not have an extra tag, the mocked class doc uses an empty tag name for the inline tag
```
Done. And sorry for my clunky English!

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+                        "", "Javadoc description",
+                        "Test extra docs group name", ""},
+                // TODO: because DefaultWorkUnitHandler does not have an extra tag, the mocked class doc uses an empty tag name for the inline tag
+                // TODO: perhaps Barclay should include a default inline tag prefix for being able to re-use between different toolkits
```
I just thought that it will be good to have tag support out-of-the-box, using the default helpdoclet. But I do not need it...

Removed TODO

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DocGenMocking.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class DocGenMocking {
```
Done.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DocGenMocking.java:
```diff
@@ -0,0 +1,73 @@
+        return mockClassDoc(javadocText, Collections.emptyMap());
+    }
+
+
```
Removed.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DocGenMocking.java:
```diff
@@ -0,0 +1,73 @@
+        return mockedClassDoc;
+    }
+
+
```
Removed.

magicDGS commented:
I addressed your comments, @cmnbroad -back to you and thanks for reviewing.

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DocGenMocks.java:
```diff
@@ -0,0 +1,71 @@
+
+    public static DocWorkUnit createDocWorkUnit(final DocWorkUnitHandler docWorkUnitHandler, final Class<?> clazz) {
+        return createDocWorkUnit(docWorkUnitHandler, clazz, emptyClassDoc());
+    }
```
These last two methods don't seem to be mock related; my preference to see the DocWorkUnit creation code inlined at the calls sites to make it clear whats going on.

magicDGS replied:
So are you suggesting to remove completely this methods? I think that they are useful, because that code is gonna be repeated all around the mocked tests...Maybe an utility class for testing the doclet?

cmnbroad replied:
Both of these (as well as emptyClassDoc and mockClassDoc(String)) are just wrappers around a single call, and to me only obscure whats being mocked and what isn't. I'd strongly prefer to see them removed and inlined now. Maybe as we develop more real tests we can factor stuff out.

magicDGS replied:
Done, although I really think that the shortcuts are useful for the future, if we keep adding more complexity to the mocked objects.

magicDGS commented on build.gradle:
```diff
@@ -59,6 +59,7 @@ dependencies {
     compile 'com.google.code.gson:gson:2.2.2'
 
     testCompile 'org.testng:testng:6.9.6'
+    testCompile 'org.mockito:mockito-core:2.8.9'
```
Done.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerUnitTest.java:
```diff
@@ -0,0 +1,88 @@
+                        .mockClassDoc("Javadoc description \nin two lines", Collections.singletonMap("", "Inline tag"))),
+                        "Inline tag"}
+        };
+    }
```
Done.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerUnitTest.java:
```diff
@@ -0,0 +1,88 @@
+        Assert.assertEquals(getNewInstance().getGroupSummaryForWorkUnit(workUnit), expectedDescription);
+    }
+
+    private static DefaultDocWorkUnitHandler getNewInstance() {
```
Done.

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DocWorkUnitTest.java:
```diff
@@ -0,0 +1,61 @@
+        workUnit.setProperty("test_property", "test_value");
+        Assert.assertEquals(workUnit.getProperty("test_property"), "test_value");
+        Assert.assertEquals(workUnit.getRootMap().size(), 1);
+    }
```
The "test_property" and "test_value" strings should be factored out into constants.

magicDGS replied:
Done.

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DocWorkUnitTest.java:
```diff
@@ -0,0 +1,61 @@
+        Assert.assertTrue(
+                DocGenMocks.createDocWorkUnit(getDefaultDocWorkUnitHandler(), TestArgumentContainer.class).getBetaFeature());
+    }
+
```
Its preferable to use a data provider for these methods wherever possible, instead of repeating the asserts for each test case.

magicDGS replied:
There are only two possibilities in this case, and in principle they shouldn't change. Anyway, I refactored...

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerUnitTest.java:
```diff
@@ -0,0 +1,89 @@
+    public static Object[][] workUnitGroupSummary() {
+        return new Object[][] {
+                {TestArgumentContainer.class, TestProgramGroup.DESCRIPTION},
+                {TestExtraDocs.class, Collections.EMPTY_MAP, ""},
```
I know you're still in the middle of this, but one has two params and one has three.

magicDGS replied:
My bad, a refactoring error. Maybe we should add a test for data providers as in HTSJDK/Picard/GATK (in a different PR).

cmnbroad commented on src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerUnitTest.java:
```diff
@@ -0,0 +1,89 @@
+    public void testGetGroupNameForWorkUnit(final Class<?> docWorkUnitClazz, final String expectedGroupName) {
+        final DocWorkUnit mockWorkUnit = DocGenMocks.createDocWorkUnit(
+                getDefaultWorkUnitHandlerInstance(), docWorkUnitClazz);
+        Assert.assertEquals(getDefaultWorkUnitHandlerInstance().getGroupNameForWorkUnit(mockWorkUnit), expectedGroupName);
```
There are several of these where there are multiple calls to getDefaultWorkUnitHandlerInstance in the same test. It think there should be one DefaultWorkUnitHandler created here, which is used to both mock the workUnit and to retrieve the groupName.

The fact that this test works as is indicates that getGroupNameForWorkUnit (and getGroupSummaryForWorkUnit ) should really be methods on DocWorkUnit, not on the handler, since it appears to not matter what handler instance is used. (I think this can up in a previous PR too). But thats for another day. For now we should fix the tests to use the same one.

magicDGS replied:
I think that it shouldn't be re-used, because we should test how a state-less WorkUnitHandler behaves.

Regarding your side note, the default handler is independent of the instance used, but the user can override the behaviour. An usage case is to document a toolkit which includes another one (e.g., Picard inside GATK). In that case, a custom handler may want to group together all the Picard programs to have them in a different page or pointing that they do not belong to the custom toolkit but to a third-party. I think that this methods belongs to the handle even if the default behaviour is instance-independent...

magicDGS replied:
No action was taken here.

cmnbroad replied:
@magicDGS The point I was making was that using a different handler to get attributes like group name than was originally used to create the work unit isn't how Barclay itself gets them, so write the test this way ? There isn't anything that says that a handler needs to be stateless - the fact that this currently works with the default handler is completely accidental. It would be quite easy to make a perfectly valid change to the default handler that would cause this test to fail, but pass if the same handler was used.

The doc should probably be more explicit about this, but in the meantime it seems much more useful, and less brittle, to have a test that ensures that this method does the right thing when used as Barclay uses it.

magicDGS replied:
Done.

magicDGS commented:
Back to you @cmnbroad!

cmnbroad commented:
@magicDGS Thanks for doing this!

magicDGS commented:
Thanks for the review, @cmnbroad!
